### DISCUSSION TITLE
Adapting color channels to match the smallest channel

### DISCUSSION THREAD

**Discussion by rabanito75** [ImageMagick/ImageMagick#3143] [Help]
Hello, i'm in the middle of digitizing countless color negatives.
I'm using dcraw to convert the camerascans into whitebalanced negatives, then use image magick to negate, mirror and save them as 8bit-lzw-tiffs.

Then i do the cropping and dust-and-scratch-retouching in Photoshop.

This leaves me with a positive image with a typical histogram similar to the attached picture below.
(the red channel is the "narrowest", the green and blue are respectively "broader")

I figured out, that if i manually adjust the output-levels to align the broader channels to the smallest instead of stretching the channels to zero using the classic "auto-color" in PS-levels or contrast-stretch in IM, the results are far superior and the process gives very consistent colors with little left to adjust. 

After adjusting the channels i then use auto-levels and a sigmoidal-contrast in IM, which gets me a result with very little to refine.

Is it possible to align / match the channels to the narrowest channel automatically in IM?

I tried to separate the channels and used Fred Weinhaus "histmatch"-script to match the green and blue channels to the red one and re-combined them. This left me with a almost perfect alignment of the channels, but the color integrity is not right, it makes the respective color channel histograms too similar. 

If i could do this process automatically, it would drastically speed up my conversion process.

Thank you very much for your answer.

![manual_levels](https://user-images.githubusercontent.com/67098233/105368313-af421b80-5c01-11eb-82be-1052501b8f8f.jpg)

**Comment by snibgo** [comment]
> I tried to separate the channels and used Fred Weinhaus "histmatch"-script to match the green and blue channels to the red one and re-combined them. This left me with a almost perfect alignment of the channels, but the color integrity is not right, it makes the respective color channel histograms too similar.

I suggest you try the "gain and bias" method. This is similar to matching histograms, but without such close matching. It would make the channels match in mean (ie lightness) and standard deviation (ie contrast). See [Gain and bias](http://im.snibgo.com/gainbias.htm).

There are alternatives. You could find the minimum and maximum of all three channels, then adjust two to match the third. However, outliers might cause problems.

**Comment by fmw42** [comment]
Try my matchimage script rather than histmatch. matchimage is similar to snibgo gain and bias method.

**Comment by rabanito75** [comment]
Thank you very much for your replies. 

I had tried the matchimage and bcmatch scripts already, they do shift the channels a bit in the right direction, but unfortunately not quiet enough as the manual process does. 

Would it be possible to get the minimum and maximum pixel values of the narrowest channel with the fx operator and apply them to the other channels with the level-operator? 

The problem though is, that the mathematical min and max values have a threshold that varies a few percent on each channel before one can actually see the histogram rising up. To evaluate the left and right edges where one can actually see the histogram in a gui would be desirable. Like clipping the values in a contrast-stretch per channel by 0.5%, but without the actual clipping...

I hope you understand what i'm trying to say, English is not my native language.. 

Thank you very much for your help.

**Comment by fmw42** [comment]
Separate channels, then 

```
Here is a special define that outputs the location of the minimum or maximum pixel of the image:

magick identify -precision 5 -define identify:locate=minimum -define identify:limit=1 image.png
```

See https://imagemagick.org/script/identify.php

You might also try my matchimage script after separating channels and before identify above

**Comment by snibgo** [comment]
The following bash script adjusts the green and blue channels to exactly match the range of the red channel.
```
magick \
  in.png \
  -set option:LEVS %[fx:minima.r*QuantumRange],%[fx:maxima.r*QuantumRange] \
  -channel G \
    -auto-level \
    +level %[LEVS] \
  -channel B \
    -auto-level \
    +level %[LEVS] \
  +channel \
  out.png
```
This may not do what you want. Please post a sample image that you want to adjust, and the result you want to achieve.

**Reply by rabanito75** [reply]
Thanks a lot for the replies, sorry for the delay.

I work on Mac OSX 10.14 with version 6.9.1-0 Q16 x86_64 2015-03-22 (sorry i didn't mention it earlier).

Do the commands also work with the convert operator or is it advisable to upgrade to version 7?

Here is a downsized sample image of a dcraw-output, cropped and inverted:
On this picture, the blue channel is the narrowest. 
The script should if possible detect the narrowest channel, then shrink the two other channels to fit the narrowest one.

![channels_to_adjust](https://user-images.githubusercontent.com/67098233/105693438-04cb4080-5f00-11eb-8f9c-a26191595504.jpg)

After manually adjusting the channels:

![adjusted_to_blue_channel](https://user-images.githubusercontent.com/67098233/105693889-92a72b80-5f00-11eb-8b91-9bc8d64e4db9.jpg)

After auto-gamma, sigmoidal-contrast 6x40% and +40% midtone color balance in the red channel:
(note, absolutely no clipping in the color channels, balanced shadows and highlights)

![final](https://user-images.githubusercontent.com/67098233/105694235-01848480-5f01-11eb-846c-d0ed8b580679.jpg)


By stretching the individual channels, a similar output is possible, but usually not without quiet some clipping of the channels, which is not desirable.

Thanks again in advance for your answers.

**Reply by fmw42** [reply]
There should be no difference between IM 6 and IM 7. But it would be a good time to upgrade to IM 7, since it has new features that are not in IM 6. If using IM 7, use magick in place of convert only, for the other methods use magick identify, magick montage, magick compare, etc.

**Comment by snibgo** [comment]
The command I showed is not valid IM v6, because "+level" can't be used with the escape "%[LEVS]".

I suggest upgrading to v7.
